tesseract/tesseract_planning/tesseract_process_managers/src/process_generators/fix_state_collision_process_generator.cpp

```diff
@@ -0,0 +1,335 @@
+    }
+  }
+
+  return true;
+}
+
+/**
+ * @brief Takes a waypoint and uses a small trajopt problem to push it out of collision if necessary
+ * @param waypoint Must be a waypoint for which getJointPosition will return a position
+ * @param input Process Input associated with waypoint. Needed for kinematics, etc.
+ * @param profile Profile containing needed params
+ * @return True if successful
+ */
+bool MoveWaypointFromCollisionTrajopt(Waypoint& waypoint,
+                                      const ProcessInput& input,
+                                      const FixStateCollisionProfile& profile)
+{
+  using namespace trajopt;
+
+  // Get position associated with waypoint
+  Eigen::VectorXd start_pos;
+  try
+  {
+    start_pos = getJointPosition(waypoint);
+  }
+  catch (std::runtime_error& e)
+  {
+    CONSOLE_BRIDGE_logError("MoveWaypointFromCollision error: %s", e.what());
+    return false;
+  }
+  std::size_t num_jnts = static_cast<std::size_t>(start_pos.size());
+
+  // Setup trajopt problem with basic info
+  ProblemConstructionInfo pci(input.tesseract);
+  pci.basic_info.n_steps = 1;
+  pci.basic_info.manip = input.manip_info.manipulator;
+  pci.basic_info.start_fixed = false;
+  pci.basic_info.use_time = false;
+
+  // Create Kinematic Object
+  pci.kin = pci.getManipulator(pci.basic_info.manip);
+
+  // Initialize trajectory to waypoint position
+  pci.init_info.type = InitInfo::GIVEN_TRAJ;
+  pci.init_info.data = tesseract_common::TrajArray(1, start_pos.size());
+  pci.init_info.data.row(0) = start_pos.transpose();
+
+  // Add constraint that position is allowed to be within a tolerance of the original position
+  {
+    Eigen::MatrixX2d limits = pci.getManipulator(pci.basic_info.manip)->getLimits().joint_limits;
+    Eigen::VectorXd range = limits.col(1).array() - limits.col(0).array();
+    Eigen::VectorXd pos_tolerance = range * profile.jiggle_factor;
+    Eigen::VectorXd neg_tolerance = range * -profile.jiggle_factor;
+
+    auto jp = std::make_shared<JointPosTermInfo>();
+    jp->coeffs = std::vector<double>(num_jnts, 1.0);
+    jp->targets = std::vector<double>(start_pos.data(), start_pos.data() + start_pos.size());
+    jp->upper_tols = std::vector<double>(pos_tolerance.data(), pos_tolerance.data() + pos_tolerance.size());
+    jp->lower_tols = std::vector<double>(neg_tolerance.data(), neg_tolerance.data() + neg_tolerance.size());
+    jp->first_step = 0;
+    jp->last_step = 0;
+    jp->name = "joint_pos";
+    jp->term_type = TT_CNT;
+    pci.cnt_infos.push_back(jp);
+  }
+
+  // Add a constraint that it must not be in collision
+  {
+    auto collision = std::make_shared<CollisionTermInfo>();
+    collision->name = "collision";
+    collision->term_type = TT_CNT;
+    collision->evaluator_type = trajopt::CollisionEvaluatorType::SINGLE_TIMESTEP;
+    collision->first_step = 0;
+    collision->last_step = 0;
+    collision->info = createSafetyMarginDataVector(pci.basic_info.n_steps, 0.025, 1);
```

Comment:
Should the contact distance be configurable?

Reply:
I can put it in the profile

Reply:
Done